### DISCUSSION TITLE
Funktion external-Datei hochladen bei 6591/6660 nicht vorhanden?

### DISCUSSION THREAD

**Discussion by AkteX** [Freetz-NG/freetz-ng#639] [Allgemein]
Bei 6591/6660 fehlt im Freetz Menü die Seite Firmware-Update. Auf dieser Seite gibt es normalerweise auch die Funktion zum Upload einer external-Datei. Leider kann der Upload der Datei deshalb nicht über das Freetz-Webif erfolgen. Gibt es diese Funktion vielleicht an einer anderen Stelle?

**Comment by Prisrak1** [comment]
reicht der interne Speicher den nicht aus?
Ich finde, bei den vielen anderen KD Boxen kann diese Funktion auch weg.

**Comment by AkteX** [comment]
External kann verwendet werden, um bestimmte Dienste auszulagern. Eine weitere Anwendung ist, dass man bestimmte Dienste, welche auf eine externe Festplatte zugreifen, erst startet, wenn diese eingebunden ist.

**Comment by Prisrak1** [comment]
war nicht die Frage von mir ..

**Comment by AkteX** [comment]
Nein der Interne Speicher reicht nicht und die external Funktion wird weiterhin benötigt. Man kann diese Datei auch im konfigurierten external Verzeichnis entpacken. Das ist halt umständlicher.

**Comment by fda77** [comment]
Um genau zu sein gibt es keine Funktion in Freetz dies Option auszubelden. Vielleicht steht ne Blume oder was zu trinken vorm Monitor?

**Reply by fda77** [reply]
Gab es zumindest nicht

**Comment by AkteX** [comment]
Laut mod.mk wird das update-Verzeichnis bei bestimmten Boxen/FW-Layouts unterdrückt. Darin ist die Datei firmware.cgi enthalten, welche u.a. für den Upload der external Datei zuständig ist.
```
$(PKG)_EXCLUDED += $(if $(or $(FREETZ_AVM_HAS_FWLAYOUT_5),$(FREETZ_AVM_HAS_FWLAYOUT_6)),usr/mww/cgi-bin/update/)
```
Könnte man die Upload Funktion evtl. auf der Seite in der Freetz Webkonfiguration einbauen, wo das external Verzeichnis festgelegt wird?

**Reply by fda77** [reply]
Das ist aber ein bisschen was anderes wie das "external-Datei hochladen" aus der Überschrift.
Für die UIMG Images ist das FW-Upload deaktiviert, da das halt was anderes ist wie bei allen anderen Methoden.
Ob das Update mit Freetz funktioniert weiss ich nich!
- ich hab nicht nachgeschaut
- ich hab kein Gerät um aus auszuprobieren
- es ist niemand in der Lage klar zu formulieren was "geht"  und was nicht. Du kannst die Zeile ja rausnehmen und berichten...

**Reply by aus-sua** [reply]
Ich glaube immer noch, s.u., dem TE geht es nur um die cgi-(Unter)-Seite "external-Datei hochladen",
er möchte kein(!) FW-Update per freetz machen...

**Comment by aus-sua** [comment]
Ich denke auch, daß hier etwas missverstanden wurde.
Hätte @AkteX gleich eine BC das WebIf hochgeladen, hätte jeder gesehen, daß es den Link zum cgi der
external-"Einspielung" nicht gibt (/cgi-bin/update/external.cgi), weil der ja in der cgi-Seite des
freetz-Firmwareupdates steckt, der eben bei bestimmten Boxen ausgeblendet wird, weil da der Upload
per freetz noch ungeklärt ist (klasse Satz... ;-(  ).

Und schlimmer, jetzt gibt es die external-"Einspielung" (... /cgi-bin/update/external.cgi) im freetz-NG nicht mehr,
wenn ich nicht explizit beim Firmware-Bau externals auswähle?
Müßte es die cgi (... /cgi-bin/update/external.cgi) nicht immer geben, unabhängig vom Firmware-Bau der externals?
Aber ja, wenn sich wirklich jemand am Link der cgi stört, weil 'eh genug Platz auf seiner Box ist (widerspricht sich
das nicht schon selbst?...), dann soll der einzelne Herr seine Option im freetz-Firmware-bau bekommen, aber bitte
nicht als Standard-Option "cgi für externals = weg ---> bei external nicht ausgewählt" sondern Standard ist, wie die
letzten 10 Jahre, die cgi für external ist immer ausgewählt...
Oder, dann sorry, habe ich da jetzt was total falsch verstanden?

**Reply by fda77** [reply]
> Und schlimmer, jetzt gibt es die external-"Einspielung" (... /cgi-bin/update/external.cgi) im freetz-NG nicht mehr,
wenn ich nicht explizit beim Firmware-Bau externals auswähle?

Was findest du daran genau schlimm? Wenn du das Paket "dtach" nicht auswählst ist es nicht im Image. Warum sollte das bei External anders sein?

**Reply by aus-sua** [reply]
Öh... wenn ich den TE richtig verstehe, geht es Ihn doch genau nur um diese Seite im
freetz-WebIf (die fehlt Ihm ja ohne Firmware-Update-freetz-cgi):
![Freetz – external-Datei Update_2022-11-10_13 17 19](https://user-images.githubusercontent.com/73757472/201104745-917b4e68-ca62-498d-821d-9b780202da4a.jpg)

Die cgi-Seite war im freetz immer da (auch ohne externals ausgewählt zu haben), der TE nutzte sie um,
um z,B. Dateien (auch externe, startbare, auf seinen Pfad einzuspielen und dann ggf. zu starten lassen
(z.B. reboot/an-abstecken USB o.ä.).
Ich sehe jetzt z.B. ein Problem darin, daß es diese cgi-Seite jetzt nur noch gibt, wenn ich
per make menuconfig wirklich explizit "externals" auswähle.
10 Jahre war diese eine poplige Seite immer da, jetzt, wenn ich sie ev. doch einmal brauche (wozu auch immer...),
dann muß ich mir eine neue Firmware erstellen, weil sich die Standard-Option geändert hat, die ich vorher nicht
einmal auswählen/verändern konnte.
Ich störe mich da nur an dem:
bis gestern --> war immer da
ab heute ---> immer weg (so lange ich es nicht immer wieder auswähle...)
Und die Begründung für die (jetzige) Standard-Optionsänderung ist nicht technischer Natur, sondern:
zuviel (!) Speicher im Flash der Box vorhanden. :-))

Wie auch immer, sooo wichtig ist mir das jetzt nicht.

**Reply by AkteX** [reply]
[>](url) Wenn ich den TE richtig verstehe, geht es Ihn doch genau nur um diese Seite im
freetz-WebIf (die fehlt Ihm ja ohne Firmware-Update-freetz-cgi)

Ja, genau darum geht es. Diese Seite ist bei mir trotz Aktivierung von "External processing" nicht vorhanden. 
![Externals Konfiguration in menuconfig](https://metarepo.tk/Bilder/Screenshot.2022-11-10.152630.png)
Ich wähle keine Pakete aus, welche von intern nach external verschoben werden sollen. Es sollen lediglich bestimmte Funktionen erst gestartet werden, wenn eine externe Festplatte angebunden ist. Diese konfiguriere ich im Freetz WebIF unter: "Folgende selbst definierten Dienste behandeln". 
![Freetz WebIF External](https://metarepo.tk/Bilder/Screenshot.2022-11-10.153316.png)
Beim Freetz make wird diese External-Datei erstellt. Die Datei externel.cgi für den Upload wird aber nicht in das Image eingebaut und ist damit im WebIF nicht verfügbar, weil der komplette Inhalt des update Verzeichnisses, inklusive external.cgi in mod.mk, unterdrückt wird.
`$(PKG)_EXCLUDED += $(if $(or $(FREETZ_AVM_HAS_FWLAYOUT_5),$(FREETZ_AVM_HAS_FWLAYOUT_6)),usr/mww/cgi-bin/update/) 
`

Es gibt weiterhin den Bereich external im Bereich Freetz -> Einstellungen. Damit sollte es auch eine Möglichkeit geben, die External-Datei über das WebIF einzuspielen. 

Die Freetz Firmware kann ich aufgrund der eigenen Signatur über das AVM WebIF einspielen.

**Comment by AkteX** [comment]
Vorschlag zur Änderung
```
diff --git a/make/pkgs/mod/files/root/usr/lib/mod/cgi/links.sh b/make/pkgs/mod/files/root/usr/lib/mod/cgi/links.sh
index 54d300c5d..4fa4be781 100644
--- a/make/pkgs/mod/files/root/usr/lib/mod/cgi/links.sh
+++ b/make/pkgs/mod/files/root/usr/lib/mod/cgi/links.sh
@@ -62,6 +62,7 @@ _cgi_location() {
                                system)    "$out" "/cgi-bin/system.cgi" ;;
                                backup)    "$out" "/cgi-bin/backup/index.cgi" ;;
                                update)    "$out" "/cgi-bin/update/firmware.cgi" ;;
+                               upload)    "$out" "/cgi-bin/update/external.cgi" ;;
                                support)   "$out" "/cgi-bin/support/index.cgi" ;;
                                freetz)    "$out" "/cgi-bin/freetz.cgi" ;;
                                extras)    "$out" "/cgi-bin/extra.cgi" ;;
diff --git a/make/pkgs/mod/files/root/usr/lib/mww/cgi/menu-new.sh b/make/pkgs/mod/files/root/usr/lib/mww/cgi/menu-new.sh
index 281888879..c769bb162 100644
--- a/make/pkgs/mod/files/root/usr/lib/mww/cgi/menu-new.sh
+++ b/make/pkgs/mod/files/root/usr/lib/mww/cgi/menu-new.sh
@@ -12,7 +12,7 @@ _cgi_submenu() {
        local id=$1
        case $id in
                status*|daemons) sub=status ;;
-               system|avmwif_*|rudi_*|firmware_*|backup_*|support_*|freetz) sub=system ;;
+               system|avmwif_*|rudi_*|firmware_*|external_*|backup_*|support_*|freetz) sub=system ;;
                *:*) sub=${id#*:}; sub="pkg:${sub%%:*}" ;;
        esac
        echo "$sub"
@@ -190,8 +190,11 @@ new_menu_prepare() {
                cat << EOF
 <li><a id="backup_restore" href="/cgi-bin/backup/index.cgi">$(lang de:"Sichern &amp; Wiederherstellen" en:"Backup &amp; restore")</a></li>
 EOF
-               [ -d /usr/mww/cgi-bin/update ] && cat << EOF
+               [ -f /usr/mww/cgi-bin/update/firmware.cgi ] && cat << EOF
 <li><a id="firmware_update" href="$(href mod update)">$(lang de:"Firmware-Update" en:"Firmware update")</a></li>
+EOF
+               [[ ! -f /usr/mww/cgi-bin/update/firmware.cgi && -f /usr/mww/cgi-bin/update/external.cgi ]] && cat << EOF
+<li><a id="external_upload" href="$(href mod upload)">$(lang de:"Upload-external" en:"Upload external")</a></li>
 EOF
                cat << EOF
 <li><a id="support_file" href="/cgi-bin/support/index.cgi">$(lang de:"Supportdatei erstellen" en:"Create support file")</a></li>
diff --git a/make/pkgs/mod/files/root/usr/mww/cgi-bin/update/external.cgi b/make/pkgs/mod/files/root/usr/mww/cgi-bin/update/external.cgi
index e64662d05..f0d0a828e 100755
--- a/make/pkgs/mod/files/root/usr/mww/cgi-bin/update/external.cgi
+++ b/make/pkgs/mod/files/root/usr/mww/cgi-bin/update/external.cgi
@@ -5,7 +5,7 @@

 . /mod/etc/conf/mod.cfg

-cgi --id=firmware_update
+cgi --id=external_upload
 cgi_begin "$(lang de:"external-Datei Update" en:"external-file update")"

 cat << EOF
diff --git a/make/pkgs/mod/mod.mk b/make/pkgs/mod/mod.mk
index e6003e76f..c82882ff7 100644
--- a/make/pkgs/mod/mod.mk
+++ b/make/pkgs/mod/mod.mk
@@ -26,7 +26,7 @@ $(PKG)_EXCLUDED += $(if $(FREETZ_REMOVE_WEBSRV),,etc/init.d/rc.websrv usr/bin/we
 $(PKG)_EXCLUDED += $(if $(FREETZ_BUSYBOX_TELNETD),,etc/init.d/rc.telnetd usr/lib/cgi-bin/conf.avm/10-telnetd.sh)
 $(PKG)_EXCLUDED += $(if $(or $(call not-y,$(EXTERNAL_ENABLED)),$(EXTERNAL_DOWNLOADER)),usr/lib/cgi-bin/mod/conf/40-external.sh etc/init.d/rc.external etc/external.pkg)
 $(PKG)_EXCLUDED += $(if $(EXTERNAL_ENABLED),,/usr/mww/cgi-bin/update/external.cgi /usr/mww/cgi-bin/update/do_external.cgi)
-$(PKG)_EXCLUDED += $(if $(FREETZ_AVM_HAS_FWLAYOUT_5),usr/mww/cgi-bin/update/)
+$(PKG)_EXCLUDED += $(if $(FREETZ_AVM_HAS_FWLAYOUT_5),usr/mww/cgi-bin/update/firmware.cgi usr/mww/cgi-bin/update/do_firmware.cgi)

 $(PKG)_EXCLUDED += $(if $(FREETZ_BUSYBOX_FEATURE_WTMP),,usr/lib/cgi-bin/mod/conf/60-utmp_wtmp.sh)
 $(PKG)_EXCLUDED += $(if $(FREETZ_TARGET_IPV6_SUPPORT),,usr/lib/cgi-bin/mod/conf/90-ipv6.sh)
```

**Comment by fda77** [comment]
Achso, bei den Puma7 / UIMIG Teilen fehlt das ganze Flashen übers Webif. Statt herumzuworkarounden könnte man auch einfach das Flashem zum laufen bringen.
Bislang ist nur "geht nischt" bekannt, genauer konnte das noch niemand eingrenzen. Ob es wirklich überhaupt Problem gibt ist auch nicht bekannt
